### PR TITLE
feat: Integration of ClickHouse storage

### PR DESCRIPTION
**Pull request by AlexTeslenko**
Goals:

ClickHouse allows quick insertion and searches over a large quantity of data. We use Clickhouse internally for search over a large corpus of embeddings.
As Docarray provide a great pythonic API to manipulate unstructured data we decided to integrate it with CLickHouse for our internal needs.
But I hope this PR will be also valuable for other engineers.

Empirically all functionality of Docarray works properly with ClickHouse. But ClickHouse is not transactional DB, therefore it does not provide atomicity of operations. 
Because of this many tests in tests/unit/array/test_advance_indexing.py do not pass for now. I'm looking for a way how to overcome this.

### CODE REVIEW

File: docarray/array/storage/clickhouse/backend.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+from typing import (
+    Dict,
+    Optional,
+    TYPE_CHECKING,
+    Union,
+    List,
+    Iterable,
+    Mapping,
+)
+if TYPE_CHECKING:
```
Comment (JoanFM) [inline docarray/array/storage/clickhouse/backend.py:12]: can you please merge all the hidden `TYPE_CHECKING` into one

File: docarray/array/storage/clickhouse/backend.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        )
+        self._config = config
+
+        super()._init_storage()
+
+        if _docs is None:
+            return
+        
+        if isinstance(_docs, Iterable):
+            self.clear()
```
Comment (JoanFM) [inline docarray/array/storage/clickhouse/backend.py:95]: lets move `self.clear()` on top of the if

File: docarray/array/storage/clickhouse/backend.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            self.extend(_docs)
+        else:
+            self.clear()
+            if isinstance(_docs, Document):
+                self.append(_docs)
+
```
Comment (JoanFM) [inline docarray/array/storage/clickhouse/backend.py:101]: shouldn't this be a private method?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: docarray/array/storage/clickhouse/backend.py (hunk 3 above)
Comment (AlexTeslenko) [reply docarray/array/storage/clickhouse/backend.py:94]: Hi @JoanFM! Thanks for the prompt response!

Both self.clear() and self.append() methods are public in the base class. I just followed a convention.

**Comment by JoanFM** [comment]
Hey @AlexTeslenko ,

I will mark it as a `draft` PR until you confirm that your implementation passes the test and is ready for review

**Comment by JoanFM** [comment]
Hello @AlexTeslenko,

Is there any progress here, any intention to continue?

**Comment by AlexTeslenko** [comment]
Hi @JoanFM! Yes. Sorry, didn't have time to work on it at work. 
I will go on my own and finish it in the next few days.

**Comment by samsja** [comment]
Hey @AlexTeslenko it seems that the PR has been hanging for a while now. Moreover we are going into a heavy redesign of the interaction with vector database in DocArray v2 : https://github.com/docarray/docarray/issues/780. Therefore it might not be pertinent to merge this PR right now. I will close it but feel free to reopen it if you still want to merge it soon rather than waiting for v2 to be out .

**Pull request closed without merging**